### PR TITLE
Fix usage items under player

### LINKED ISSUE #4001: Item usage detection bug
### Before creating an issue, please ensure:
- [x] This is a bug in the software that resides in this repository, and not a
      support matter (use https://otland.net/forums/support.16/ for support)
- [x] This issue is reproducible without changes to the C++ code in this repository

### Steps to reproduce (include any configuration/script required to reproduce)
1. Put any item bellow your character
2. Try using it...

### Expected behaviour
Item used successfully 

### Actual behaviour
Item not used, but your character was "used" instead

### More info
https://github.com/Zbizu/forgottenserver/commit/36c45f2dcc8f7e9826eafb65e7d0efe477a2dc60#diff-0bb6c7b942a6c15255f935db679c066dfce4d4114fdb99698c1e1a09b7f5a98aL2056
https://github.com/Zbizu/forgottenserver/commit/ac4e65522c0c948b31c47e73bf4ab044347ccd6a
https://github.com/Zbizu/forgottenserver/commit/018505efc4f99ca7e22e6941a545dbe983e587b5
https://github.com/Zbizu/forgottenserver/commit/78fa3ee5dd9f792b3db619c82df83ed7dc435b66
https://github.com/Zbizu/forgottenserver/commit/f199e14d31fa540bbb986b70f4385ff2c51651e9
https://github.com/Zbizu/forgottenserver/commit/d4a41c5ab9240db14c5422d45ec1acbc83098a6d

### PR DESCRIPTION
**Pull request by ArturKnopik**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
Fix usage items under player
closes #4001

### How to test
1. Put item under player, for example mana potion
2. Use item
-------------------------------------------------------
1. Stay under door
2. try close the door

Code author: @Zbizu

### CODE REVIEW

**Comment by Zbizu** [comment]
correct, original diff (private repo):
```diff
diff --git a/src/tile.cpp b/src/tile.cpp
index b748de0c2..5b42a8b8e 100644
--- a/src/tile.cpp
+++ b/src/tile.cpp
@@ -1606,12 +1606,31 @@ bool Tile::isMoveableBlocking() const
 Item* Tile::getUseItem(int32_t index) const
 {
 	const TileItemVector* items = getItemList();
+
+	// no items, get ground
 	if (!items || items->size() == 0) {
 		return ground;
 	}
 
+	// try getting thing by index
 	if (Thing* thing = getThing(index)) {
-		return thing->getItem();
+		Item* thingItem = thing->getItem();
+		if (thingItem) {
+			return thingItem;
+		}
+	}
+
+	// try getting top usable item
+	Item* topDownItem = getTopDownItem();
+	if (topDownItem) {
+		return topDownItem;
+	}
+
+	// try getting door
+	for (auto it = items->rbegin(), end = items->rend(); it != end; ++it) {
+		if ((*it)->getDoor()) {
+			return (*it)->getDoor();
+		}
 	}
 
 	return nullptr;
```

**Review by ranisalt — APPROVED** [review]

**Review by EPuncker — APPROVED** [review]

File: src/tile.cpp [hunk 1]
```diff
@@ -1606,27 +1606,31 @@ Item* Tile::getUseItem(int32_t index) const
...
 
+	// try getting thing by index
 	if (Thing* thing = getThing(index)) {
 		Item* thingItem = thing->getItem();
 		if (thingItem) {
 			return thingItem;
 		}
 	}
 
+	// try getting top usable item
```
Comment (Zbizu) [inline src/tile.cpp:1622]: top movable*

File: src/tile.cpp [hunk 2]
```diff
@@ -1606,27 +1606,31 @@ Item* Tile::getUseItem(int32_t index) const
...
+	// try getting door
 	for (auto it = items->rbegin(), end = items->rend(); it != end; ++it) {
 		if ((*it)->getDoor()) {
-			return (*it)->getItem();
+			return (*it)->getDoor();
 		}
 	}
 
-	return *items->begin();
+	return nullptr;
```
Comment (Zbizu) [inline src/tile.cpp:1635]: This needs to stay. It returns top immovable item.

**Review by Zbizu — CHANGES_REQUESTED** [review]

**Review by EPuncker — APPROVED** [review]

**Comment by kaleohanopahala** [comment]
Hmm would be:
It's prevent double interaction. 
Safe returns.

```
Item* Tile::getUseItem(int32_t index) const
{
	const TileItemVector* items = getItemList();

	// no items, return ground
	if (!items || items->empty()) {
		return ground;
	}

	// try getting thing by index if it's within bounds
	if (index >= 0 && index < static_cast<int32_t>(items->size())) {
		if (Thing* thing = getThing(index)) {
			Item* thingItem = thing->getItem();
			if (thingItem) {
				return thingItem;
			}
		}
	}

	// try getting top movable item
	Item* topDownItem = getTopDownItem();
	if (topDownItem) {
		return topDownItem;
	}

	// try getting door item
	for (auto it = items->rbegin(), end = items->rend(); it != end; ++it) {
		if ((*it)->getDoor()) {
			return (*it)->getItem();
		}
	}

	// fallback to first item in the list if items are present
	return !items->empty() ? *items->begin() : nullptr;
}
```

**Pull request merged**